File: lib/src/main/java/com/llama4j/FloatTensor.java
package com.llama4j;

import jdk.incubator.vector.FloatVector;
import jdk.incubator.vector.VectorShape;
import jdk.incubator.vector.VectorSpecies;

import java.lang.foreign.MemorySegment;
import java.lang.reflect.Field;
import java.util.Arrays;

abstract class FloatTensor {
    static final int VECTOR_BIT_SIZE = Integer.getInteger("llama.VectorBitSize", VectorShape.preferredShape().vectorBitSize());
    static final boolean USE_VECTOR_API = VECTOR_BIT_SIZE != 0;

    static final VectorSpecies<Float> F_SPECIES;
    static final VectorSpecies<Integer> I_SPECIES;
    static final VectorSpecies<Short> S_SPECIES_HALF;

    static {
        if (USE_VECTOR_API) {
            F_SPECIES = VectorShape.forBitSize(VECTOR_BIT_SIZE).withLanes(float.class);
            I_SPECIES = F_SPECIES.withLanes(int.class);
            S_SPECIES_HALF = VectorShape.forBitSize(F_SPECIES.vectorBitSize() / 2).withLanes(short.class);
            assert F_SPECIES.length() == S_SPECIES_HALF.length();
        } else {
            F_SPECIES = null;
            I_SPECIES = null;
            S_SPECIES_HALF = null;
        }
    }

    static final sun.misc.Unsafe UNSAFE;

    static {
        try {
            Field f = sun.misc.Unsafe.class.getDeclaredField("theUnsafe");
            f.setAccessible(true);
            UNSAFE = (sun.misc.Unsafe) f.get(null);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }

    static short readShort(MemorySegment memorySegment, long offset) {
        return UNSAFE.getShort(memorySegment.address() + offset);
    }

    static float readFloat16(MemorySegment memorySegment, long offset) {
        return Float.float16ToFloat(readShort(memorySegment, offset));
    }

    static byte readByte(MemorySegment memorySegment, long offset) {
        return UNSAFE.getByte(memorySegment.address() + offset);
    }

    static float readFloat(MemorySegment memorySegment, long offset) {
        return UNSAFE.getFloat(memorySegment.address() + offset);
    }

    abstract long size();

    abstract float getFloat(long index);

    abstract void setFloat(int index, float value);

    abstract FloatVector getFloatVector(VectorSpecies<Float> species, int offset);

    abstract GGMLType type();

    public static int numberOfElements(int... dimensions) {
        assert Arrays.stream(dimensions).allMatch(i -> i > 0);
        return Arrays.stream(dimensions).reduce(Math::multiplyExact).orElseThrow();
    }

    public static long numberOfElementsLong(int... dimensions) {
        long result = 1;
        for (int d : dimensions) {
            assert d > 0;
            result = Math.multiplyExact(result, d);
        }
        return result;
    }

    static float scalarDot(FloatTensor thiz, int thisOffset, FloatTensor that, int thatOffset, int size) {
        float result = 0f;
        for (int j = 0; j < size; j++) {
            result += thiz.getFloat(thisOffset + j) * that.getFloat(thatOffset + j);
        }
        return result;
    }

    float dot(int thisOffset, FloatTensor that, int thatOffset, int size) {
        return scalarDot(this, thisOffset, that, thatOffset, size);
    }

    void matmul(FloatTensor that, FloatTensor out, int dim0, int dim1) {
        Parallel.parallelFor(0, dim0, i -> out.setFloat(i, dot(i * dim1, that, 0, dim1)));
    }

    // matmul with offset into this tensor (for expert weight slicing in 3D tensors)
    void matmul(FloatTensor that, FloatTensor out, int dim0, int dim1, int thisOffset) {
        Parallel.parallelFor(0, dim0, i -> out.setFloat(i, dot(thisOffset + i * dim1, that, 0, dim1)));
    }

    @FunctionalInterface
    interface AggregateFunction {
        float apply(float acc, float value);
    }

    float reduce(int thisOffset, int size, float seed, AggregateFunction reduce) {
        float result = seed;
        for (int i = 0; i < size; ++i) {
            result = reduce.apply(result, getFloat(thisOffset + i));
        }
        return result;
    }

    float sum(int thisOffset, int size) {
        return reduce(thisOffset, size, 0f, Float::sum);
    }

    float max(int thisOffset, int size) {
        return reduce(thisOffset, size, Float.NEGATIVE_INFINITY, Float::max);
    }

    void copyTo(int thisOffset, FloatTensor that, int thatOffset, int size) {
        that.mapWithIndexInPlace(thatOffset, size, (value, index) -> this.getFloat(index - thatOffset + thisOffset));
    }

    int argmax(int thisOffset, int size) {
        assert size > 0;
        int maxIndex = thisOffset;
        float maxValue = this.getFloat(maxIndex);
        int endIndex = thisOffset + size;
        for (int i = thisOffset; i < endIndex; ++i) {
            float f = this.getFloat(i);
            if (f > maxValue) {
                maxValue = f;
                maxIndex = i;
            }
        }
        return maxIndex;
    }

    int argmax() {
        return argmax(0, Math.toIntExact(size()));
    }

    @FunctionalInterface
    interface MapFunction {
        float apply(float value);
    }

    @FunctionalInterface
    interface MapWithIndexFunction {
        float apply(float value, int index);
    }

    FloatTensor mapInPlace(int thisOffset, int size, MapFunction mapFunction) {
        int endIndex = thisOffset + size;
        for (int i = thisOffset; i < endIndex; ++i) {
            setFloat(i, mapFunction.apply(getFloat(i)));
        }
        return this;
    }

    FloatTensor mapInPlace(MapFunction mapFunction) {
        return mapInPlace(0, Math.toIntExact(size()), mapFunction);
    }

    FloatTensor mapWithIndexInPlace(int thisOffset, int size, FloatTensor.MapWithIndexFunction mapWithIndexFunction) {
        int endOffset = thisOffset + size;
        for (int i = thisOffset; i < endOffset; ++i) {
            setFloat(i, mapWithIndexFunction.apply(getFloat(i), i));
        }
        return this;
    }

    FloatTensor addInPlace(int thisOffset, FloatTensor that, int thatOffset, int size) {
        return mapWithIndexInPlace(thisOffset, size, (value, index) -> value + that.getFloat(index - thisOffset + thatOffset));
    }

    FloatTensor addInPlace(FloatTensor that) {
        return addInPlace(0, that, 0, Math.toIntExact(size()));
    }

    FloatTensor multiplyInPlace(int thisOffset, FloatTensor that, int thatOffset, int size) {
        return mapWithIndexInPlace(thisOffset, size, (value, index) -> value * that.getFloat(index - thisOffset + thatOffset));
    }

    FloatTensor divideInPlace(int thisOffset, int size, float value) {
        return mapInPlace(thisOffset, size, f -> f / value);
    }

    FloatTensor fillInPlace(int thisOffset, int size, float value) {
        return mapInPlace(thisOffset, size, unused -> value);
    }

    FloatTensor softmaxInPlace(int thisOffset, int size) {
        float maxVal = max(thisOffset, size);
        mapInPlace(thisOffset, size, f -> (float) Math.exp(f - maxVal));
        float sum = sum(thisOffset, size);
        return divideInPlace(thisOffset, size, sum);
    }

    FloatTensor saxpyInPlace(int thisOffset, FloatTensor that, int thatOffset, int size, float a) {
        for (int i = 0; i < size; ++i) {
            setFloat(thisOffset + i, a * that.getFloat(thatOffset + i) + this.getFloat(thisOffset + i));
        }
        return this;
    }
}

File: lib/src/commonMain/kotlin/io/github/lbenedetto/internal/floattensor/FloatTensor.kt
package io.github.lbenedetto.internal.floattensor

import io.github.lbenedetto.internal.util.Math
import io.github.lbenedetto.internal.util.assert
import kotlin.math.exp
import kotlin.math.max

interface FloatTensor {
  val size: Long

  fun getFloat(index: Long): Float

  fun setFloat(index: Int, value: Float)

  fun dot(thisOffset: Int, that: FloatTensor, thatOffset: Int, size: Int): Float {
    return scalarDot(this, thisOffset, that, thatOffset, size)
  }

  fun matmul(that: FloatTensor, out: FloatTensor, dim0: Int, dim1: Int) {
    parallelFor(0, dim0) { out.setFloat(it, dot(it * dim1, that, 0, dim1)) }
  }

  // matmul with offset into this tensor (for expert weight slicing in 3D tensors)
  fun matmul(that: FloatTensor, out: FloatTensor, dim0: Int, dim1: Int, thisOffset: Int) {
    parallelFor(0, dim0) { out.setFloat(it, dot(thisOffset + it * dim1, that, 0, dim1)) }
  }

  fun reduce(thisOffset: Int, size: Int, seed: Float, reduce: (Float, Float) -> Float): Float {
    var result = seed
    for (i in 0..<size) {
      result = reduce(result, getFloat((thisOffset + i).toLong()))
    }
    return result
  }

  fun sum(thisOffset: Int, size: Int): Float {
    return reduce(thisOffset, size, 0f, Float::plus)
  }

  fun max(thisOffset: Int, size: Int): Float {
    return reduce(thisOffset, size, Float.NEGATIVE_INFINITY) { a, b -> max(a, b) }
  }

  fun copyTo(thisOffset: Int, that: FloatTensor, thatOffset: Int, size: Int) {
    that.mapWithIndexInPlace(
      thatOffset,
      size
    ) { _, index -> this.getFloat((index - thatOffset + thisOffset).toLong()) }
  }

  fun argmax(thisOffset: Int, size: Int): Int {
    assert(size > 0)
    var maxIndex = thisOffset
    var maxValue = this.getFloat(maxIndex.toLong())
    val endIndex = thisOffset + size
    for (i in thisOffset..<endIndex) {
      val f = this.getFloat(i.toLong())
      if (f > maxValue) {
        maxValue = f
        maxIndex = i
      }
    }
    return maxIndex
  }

  fun argmax(): Int {
    return argmax(0, Math.toIntExact(size))
  }

  fun mapInPlace(thisOffset: Int, size: Int, mapFunction: (Float) -> Float): FloatTensor {
    val endIndex = thisOffset + size
    for (i in thisOffset..<endIndex) {
      setFloat(i, mapFunction(getFloat(i.toLong())))
    }
    return this
  }

  fun mapInPlace(mapFunction: (Float) -> Float): FloatTensor {
    return mapInPlace(0, Math.toIntExact(size), mapFunction)
  }

  fun mapWithIndexInPlace(thisOffset: Int, size: Int, mapWithIndexFunction: (Float, Int) -> Float): FloatTensor {
    val endOffset = thisOffset + size
    for (i in thisOffset..<endOffset) {
      setFloat(i, mapWithIndexFunction(getFloat(i.toLong()), i))
    }
    return this
  }

  fun addInPlace(thisOffset: Int, that: FloatTensor, thatOffset: Int, size: Int): FloatTensor {
    return mapWithIndexInPlace(
      thisOffset,
      size
    ) { value, index -> value + that.getFloat((index - thisOffset + thatOffset).toLong()) }
  }

  fun addInPlace(that: FloatTensor): FloatTensor {
    return addInPlace(0, that, 0, Math.toIntExact(size))
  }

  fun multiplyInPlace(thisOffset: Int, that: FloatTensor, thatOffset: Int, size: Int): FloatTensor {
    return mapWithIndexInPlace(
      thisOffset,
      size
    ) { value, index -> value * that.getFloat((index - thisOffset + thatOffset).toLong()) }
  }

  fun divideInPlace(thisOffset: Int, size: Int, value: Float): FloatTensor {
    return mapInPlace(thisOffset, size) { it / value }
  }

  open fun fillInPlace(thisOffset: Int, size: Int, value: Float): FloatTensor {
    return mapInPlace(thisOffset, size) { value }
  }

  fun softmaxInPlace(thisOffset: Int, size: Int): FloatTensor {
    val maxVal = max(thisOffset, size)
    mapInPlace(thisOffset, size) { exp((it - maxVal).toDouble()).toFloat() }
    val sum = sum(thisOffset, size)
    return divideInPlace(thisOffset, size, sum)
  }

  fun saxpyInPlace(thisOffset: Int, that: FloatTensor, thatOffset: Int, size: Int, a: Float): FloatTensor {
    for (i in 0..<size) {
      setFloat(thisOffset + i, a * that.getFloat((thatOffset + i).toLong()) + this.getFloat((thisOffset + i).toLong()))
    }
    return this
  }

  companion object {

    fun Byte.toUnsignedInt(): Int = toInt() and 0xFF

    fun Short.toUnsignedInt(): Int = toInt() and 0xFFFF

    fun numberOfElements(vararg dimensions: Int): Int {
      assert(dimensions.all { it > 0 })
      return dimensions.reduce(Math::multiplyExact)
    }

    fun numberOfElementsLong(vararg dimensions: Int): Long {
      var result: Long = 1
      for (d in dimensions) {
        assert(d > 0)
        result = Math.multiplyExact(result, d.toLong())
      }
      return result
    }

    fun scalarDot(thiz: FloatTensor, thisOffset: Int, that: FloatTensor, thatOffset: Int, size: Int): Float {
      var result = 0f
      for (j in 0..<size) {
        result += thiz.getFloat((thisOffset + j).toLong()) * that.getFloat((thatOffset + j).toLong())
      }
      return result
    }
  }
}
